Use non-default device when tesing MI300A

tcclevenger opened the pull request:
Anecdotally, when I've noticed runtimes slowing on Odyssey, almost all the work in on device 0. Switching to non-default device could then speed up testing at times when many user are on the machine, since there is no scheduling.

Also, a test using non-default device would improve coverage in general, since we've not caught bugs in the past with the threading test because all testing was on default device. This is not crucial since we now have multi-gpu testing, but could be a minor benefit.

tcclevenger committed:
Use non-default device when tesing MI300A

The machine used is a shared resource with no scheduling, meaning that
often we are running at the same time as others. Most of the usage is on
device 0, so using a different device could help runtimes on occasions.

Signed-off-by: tcclevenger <tccleve@sandia.gov>

tcclevenger commented:
@masterleinad @dalg24 Let me know if y'all think this is worth it or not. I tested by running `Kokkos_CoreUnitTest_HIP` continually on dev0 and then timed running another instance on dev0 and dev3, and the time went from 3min10sec to 2min40sec. I think that is better than the worst case scenario since in `Kokkos_CoreUnitTest_HIP` the GPUs aren't constantly being saturated (like could happen with a random user).

masterleinad reviewed (approved):
I was thinking about doing the same at some point.

tcclevenger committed:
Relax Device ID test

Now, device ID for testing is not necessarily 0. Check that it is in range 0 <= dev_id < num_devices.

Signed-off-by: tcclevenger <tccleve@sandia.gov>
